adding tests for create and fund account

jeesunikim opened the pull request:
**Summary:**
- Added tests for: `create account`, `fund account`, `create muxed account`, `parse muxed account`

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

jeesunikim commented on tests/fundAccountPage.test.ts:
```diff
@@ -0,0 +1,148 @@
+    await expect(getLumenButton).toBeEnabled();
+
+    // Mock the friendbot api call
+    await page.route("**/*", async (route, request) => {
```
mocking friendbot api call

quietbits replied:
Nice! Could we use full route string here `*/**/?addr=GDVOT2ALMUF3G54RBHNJUEV6LOAZCQQCARHEVNUPKGMVPWFC4PFN33QR` and remove the `if` check?

jeesunikim commented on tests/fundAccountPage.test.ts:
```diff
@@ -0,0 +1,148 @@
+    await expect(publicKeyInput).toHaveAttribute("aria-invalid", "false");
+    await expect(getLumenButton).toBeEnabled();
+
+    await getLumenButton.click();
```
I could also mock the api call to return an error vs. clicking the button twice

quietbits replied:
Yes, let's mock an error response to control this better.

quietbits commented on tests/createAccountPage.test.ts:
```diff
@@ -0,0 +1,39 @@
+      .getByRole("button", { name: "Fund account with Friendbot" })
+      .click();
+
+    await page.goto("http://localhost:3000/account/fund");
```
The Fund account button should take us to the next page, right? I think we want to check if that happened using [`toHaveURL` method](https://playwright.dev/docs/api/class-pageassertions#page-assertions-to-have-url).

quietbits commented on tests/createMuxedAccountPage.test.ts:
```diff
@@ -0,0 +1,63 @@
+
+    await createButton.click();
+
+    await expect(page.getByTestId("createAccount-success")).toBeVisible();
```
We could also check that the correct muxed account address was created.

jeesunikim replied:
I can use mainnet muxed account value since testnet one will reset.

quietbits commented on tests/fundAccountPage.test.ts:
```diff
@@ -0,0 +1,148 @@
+  test("Shows futurenet network in the title if I change my network to futurenet", async ({
+    page,
+  }) => {
+    // Click network selector dropdown button
+    await page.getByTestId("networkSelector-button").click();
+    await expect(page.getByTestId("networkSelector-dropdown")).toBeVisible();
+
+    // Select Futurenet in the dropdown list
+    await page
+      .getByTestId("networkSelector-dropdown")
+      .getByTestId("networkSelector-option")
+      .filter({ has: page.getByText("Futurenet") })
+      .click();
+
+    // Network Submit button
+    const submitButton = page.getByTestId("networkSelector-submit-button");
+
+    // Select 'Futurenet' in the network dropdown list
+    await expect(submitButton).toHaveText("Switch to Futurenet");
+    await expect(submitButton).toBeEnabled();
+
+    // Click 'Switch to Futurenet' button
+    await submitButton.click();
+
+    await expect(page.getByTestId("networkSelector-button")).toHaveText(
+      "Futurenet",
+    );
```
We could simplify the network selection by setting a URL. We test Network selection separately, so we can assume that it's working here.

```suggestion
    await page.goto("http://localhost:3000/account/muxed-create?||=network$id=futurenet&label=Futurenet&horizonUrl=https:////horizon-futurenet.stellar.org&rpcUrl=https:////rpc-futurenet.stellar.org&passphrase=Test%20SDF%20Future%20Network%20/;%20October%202022;;")
`

quietbits commented on tests/fundAccountPage.test.ts:
```diff
@@ -0,0 +1,148 @@
+      "Friendbot: fund a futurenet network account",
+    );
+  });
+
```
Should we also test that friendbot is disabled for mainnet?

quietbits commented on tests/fundAccountPage.test.ts:
```diff
@@ -0,0 +1,148 @@
+      .getByTestId("fundAccount-buttons")
+      .getByText("Get lumens");
+
+    // Type in an invalid string in 'Public Key' input field
```
```suggestion
    // Type in an valid string in 'Public Key' input field
```

quietbits commented on tests/fundAccountPage.test.ts:
```diff
@@ -0,0 +1,148 @@
+      .getByTestId("fundAccount-buttons")
+      .getByText("Get lumens");
+
+    // Type in an invalid string in 'Public Key' input field
```
```suggestion
    // Type in an valid string in 'Public Key' input field
```

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

jeesunikim commented on src/app/(sidebar)/account/fund/page.tsx:
```diff
@@ -30,6 +35,15 @@ export default function FundAccount() {
     }
   }, [isError, isSuccess]);
 
+  if (!IS_TESTING_NETWORK) {
```
I also tried out with adding a [middleware.ts](https://nextjs.org/docs/app/building-your-application/routing/middleware) to reroute to `404` if `nextUrl` includes `mainnet` in params. However, it doesn't always work in all cases. ex. user directly types in `/account/fund` in a browser which doesn't populate params until after user lands on the page.

another method i tried is `router.push` on the client side but that produced a glitch which wasn't a good ux.

so I came up with this.

jeesunikim replied:
I'll update the copy in the next pr when I make the overall changes following the updates

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

quietbits commented:
Not related to tests, the generated public address wasn't cleared when I switched from Testnet to Mainnet.

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

jeesunikim commented:
> Not related to tests, the generated public address wasn't cleared when I switched from Testnet to Mainnet.

added this feedback in [the new issue ticket](https://github.com/stellar/laboratory/issues/819) created by Jane

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

quietbits commented on src/app/(sidebar)/account/layout.tsx:
```diff
@@ -4,33 +4,50 @@ import React from "react";
+  {
+    route: Routes.ACCOUNT_FUND,
+    label: "Fund Account",
+    isTestnetOnly: true,
```
We won't need this extra logic for the new update where we're not hiding the nav, right? Just a note for the next PR. :nerd_face:

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://laboratory-pr808.previews.kube001.services.stellar-ops.com/
